### PR TITLE
feat(core, webserver): add /v1/models api

### PR DESCRIPTION
**Pull request by antimonyGu**
ticket #3097
- Test:
**toml file content**
```
~/.tabby$ cat config.toml 
# Completion model
[model.completion.http]
kind = "llama.cpp/completion"
api_endpoint = "http://localhost:8888"
prompt_template = "<PRE> {prefix} <SUF>{suffix} <MID>"  # Example prompt template for the CodeLlama model series.
supported_models = [ "red", "yellow", "green" ]
model_name = "wsxiaoys/StarCoder-1B"

# Chat model
[model.chat.http]
kind = "openai/chat"
api_endpoint = "http://localhost:8888"
supported_models = [ "red", "yellow", "green" ]
model_name = "Qwen2-1.5B-Instruct"

# Embedding model
[model.embedding.http]
kind = "llama.cpp/embedding"
api_endpoint = "http://localhost:8888"
supported_models = [ "red", "yellow", "green" ]
model_name = "Qwen2-1.5B-Instruct"

```

**curl test result**

```
curl -L 'http://localhost:8080/v1beta/models' -H 'Accept: application/json' -H 'Content-Type: application/json' -H 'Authorization: Bearer auth_02e5f4bf9abd4754a281c9137a0ddc12'
{"completion":["red","yellow","green"],"chat":["red","yellow","green"]}
```

### CODE REVIEW

File: crates/tabby/src/routes/server_setting.rs
Comment (wsxiaoys) [inline crates/tabby/src/routes/server_setting.rs:1]: Please create new file for this endpoint.

File: crates/tabby/src/serve.rs [hunk 1]
```diff
@@ -313,7 +313,9 @@ async fn api_router(
     }
 
     let server_setting_router =
-        Router::new().route("/v1beta/server_setting", routing::get(routes::setting));
+        Router::new()
+            .route("/v1beta/server_setting", routing::get(routes::setting))
+            .route("/v1/models", routing::get(routes::models));
```
Comment (wsxiaoys) [inline crates/tabby/src/serve.rs:318]: move this to line router created in line 249

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

**Comment by antimonyGu** [comment]
Fixing comment and will update.

File: crates/tabby/src/routes/supported_models.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    tag = "v1beta",
+    operation_id = "config",
+    responses(
+        (status = 200, description = "Success", body = ServerSetting, content_type = "application/json"),
+    ),
+    security(
+        ("token" = [])
+    )
+)]
+pub async fn models() -> Json<Vec<String>> {
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:16]: return an json object instead, e.g

```
{
  completion: [...],
  chat: [...],
}
```

File: crates/tabby/src/routes/supported_models.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    let mut http_model_configs: Vec<String> = Vec::new();
+
+    match models.embedding {
+        tabby_common::config::ModelConfig::Http(http_config) => {
+            http_model_configs.extend(http_config.supported_models.unwrap());
+        }
+        tabby_common::config::ModelConfig::Local(_) => {
+            return Json(http_model_configs)
+        }
+    }
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:27]: no need to append embedding models

**Review by wsxiaoys — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: crates/tabby/src/routes/mod.rs [hunk 4]
```diff
@@ -57,9 +57,11 @@ mod completions;
 mod events;
 mod health;
 mod server_setting;
+mod supported_models;
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/mod.rs:60]:
```suggestion
mod models;
```

File: crates/tabby/src/routes/supported_models.rs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+use axum::Json;
+use serde::{Deserialize, Serialize};
+use tabby_common::config::Config;
+use utoipa::ToSchema;
+
+#[derive(Serialize, Deserialize, ToSchema, Clone, Debug)]
+pub struct SupportedModel {
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:7]:
```suggestion
pub struct ModelInfo {
```

File: crates/tabby/src/routes/supported_models.rs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    responses(
+        (status = 200, description = "Success", body = ServerSetting, content_type = "application/json"),
+    ),
+    security(
+        ("token" = [])
+    )
+)]
+pub async fn models() -> Json<SupportedModel> {
+    let models: tabby_common::config::ModelConfigGroup =
+        Config::load().expect("Config file should be exist").model;
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:26]: Config shall be passed as a state from `serve.rs`.

See https://github.com/TabbyML/tabby/blob/80f34f9d8f7e1fac3cb27aa193456daed265347c/crates/tabby/src/routes/health.rs#L18 as an example (HealthState).

File: crates/tabby/src/routes/supported_models.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    let mut http_model_configs: SupportedModel = SupportedModel {
+        completion: Vec::new(),
+        chat: Vec::new(),
+    };
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:30]:
```suggestion
    let mut completion = None;
    let mut chat = None;
```

File: crates/tabby/src/routes/supported_models.rs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    if let Some(tabby_common::config::ModelConfig::Http(completion_http_config)) = models.completion
+    {
+        if let Some(not_none_supported_models) = completion_http_config.supported_models {
+            http_model_configs
+                .completion
+                .extend(not_none_supported_models);
+        }
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:38]:
```suggestion
        if let Some(models) = completion_http_config.supported_models {
            completion = Some(models.clone());
        }
```

File: crates/tabby/src/routes/supported_models.rs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+use axum::Json;
+use serde::{Deserialize, Serialize};
+use tabby_common::config::Config;
+use utoipa::ToSchema;
+
+#[derive(Serialize, Deserialize, ToSchema, Clone, Debug)]
+pub struct SupportedModel {
+    completion: Vec<String>,
+    chat: Vec<String>,
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/supported_models.rs:9]:
```suggestion
    completion: Option<Vec<String>>,
    chat: Option<Vec<String>>,
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: crates/tabby/src/routes/models.rs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+use axum::{extract::State, Json};
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+#[derive(Serialize, Deserialize, ToSchema, Clone, Debug)]
+pub struct ModelInfo {
+    completion: Option<Vec<String>>,
+    chat: Option<Vec<String>>,
+}
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/models.rs:11]: implement `From<Config>` for `ModelInfo`

File: crates/tabby/src/routes/models.rs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    tag = "v1beta",
+    operation_id = "config",
+    responses(
+        (status = 200, description = "Success", body = ServerSetting, content_type = "application/json"),
+    ),
+    security(
+        ("token" = [])
+    )
+)]
+pub async fn models(State(state): State<Arc<tabby_common::config::Config>>) -> Json<ModelInfo> {
```
Comment (wsxiaoys) [inline crates/tabby/src/routes/models.rs:25]:
```suggestion
pub async fn models(State(state): State<Arc<ModelInfo>) -> Json<ModelInfo> {
```

File: crates/tabby/src/serve.rs [hunk 12]
```diff
@@ -257,6 +257,8 @@ async fn api_router(
                 "/v1/health",
                 routing::get(routes::health).with_state(health_state),
             )
+            .route("/v1beta/models", routing::get(routes::models))
+            .with_state(Arc::new(config.clone()))
```
Comment (wsxiaoys) [inline crates/tabby/src/serve.rs:261]: convert config to ModelInfo using `.into`

**Review by wsxiaoys — APPROVED** [review]
Otherwise LGTM

**Pull request merged**